components/emoji_picker/emoji_picker.jsx

```diff
@@ -404,12 +396,96 @@ export default class EmojiPicker extends React.Component {
                     className='emoji-picker__search'
                     type='text'
                     onChange={this.handleFilterChange}
+                    onKeyDown={this.handleKeyDown}
                     placeholder={Utils.localizeMessage('emoji_picker.search', 'search')}
                 />
             </div>
         );
     }
 
+    emojiCurrentResults() {
```

Comment:
I'm still a bit stuck on how deeply this function is nested and that it's all done inline in the JSX, but I think I've figured out how to simplify it. Since the main thing we care about carrying through the functions is the number of emojis loaded so far, what if we did it like
```jsx
emojiCurrentResults = () => {
    const categories = ...;
    
    let numEmojisLoaded = 0; // I think starting this at 0 and then incrementing it afterwards makes a bit more sense
    const components = [];
    for (let i = 0; i < categories.length; i++) {
        const category = this.getCategoriesByKey(categories[i]);
        const emojis = ...;

        const items = this.renderEmojisInCategory(i, emojis, numEmojisLoaded);
        numEmojisLoaded += items.length;

        return (
            <EmojiPickerCategorySection ...>
                {items}
            </EmojiPickerCategorySection>
        );
    }

    return ...;
};

renderEmojisInCategory = (categoryIndex, emojis, numEmojisLoaded) => {
    const {cursor, filter} = this.state;

    return emojis.map((emoji, index) => { // This could even be broken out as well if we wanted
        if (numEmojisLoaded + index >= this.state.emojisToShow) {
            let ref;
            if (numEmojisLoaded + index === this.state.emojisToShow) {
                ref = this.lastVisibleEmojisRef;
            }

            return ...;
        }

        return ...
    };
};
```

Reply:
I like the way you think. I implemented something very similar to this and it works great. I am going to go over it more though and see if I can make it follow react convention a little better. Break things out and pass in some props. Add some proper componentWillUpdate structure.

👍